### ISSUE TITLE
Add native driver for ClickHouse

### ISSUE THREAD

**Issue by hanson-young** [sfu-db/connector-x#508]
#### Try Discussion?

Thank you for your outstanding contribution. I have read your paper and think it is great, but I suggest not using MySQL to support the connection to ClickHouse. It is better to use the native ClickHouse driver. 

This is because the upstream library [r2d2-mysql ](https://github.com/outersky/r2d2-mysql)depends on https://github.com/blackbeam/rust-mysql-simple/tree/master, which performs some handshake operations when connecting to the database server, such as querying max_allowed_packet. Some of these commands are not supported by ClickHouse, which may lead to unsuccessful connections and make it difficult for users to get started directly.
![image](https://github.com/sfu-db/connector-x/assets/11735179/8660aef6-548f-4092-8e8b-90417fb20629)

![image](https://github.com/sfu-db/connector-x/assets/11735179/b5c68c61-69c1-4d46-9297-c9f3baea3fca)

#### Other

Feel free to ask here if you think it's more suitable.

**Comment by wangxiaoying** [comment]
Hi @hanson-young , thank you for starting this discussion! Indeed, we do observe some issues using mysql to connect to clickhouse both in terms of functionality (e.g., authentication) and performance. And it will be ideal if we can add a native clickhouse driver in connectorx. (I saw there are several clickhouse drivers in rust, not sure which one is the best choice.)

We may not have time to support this in this month. Please feel free to contribute and let me know if you have any questions!

**Comment by Spar9a** [comment]
Hi @wangxiaoying, I think it could also be related to my problem.

My team and I using clickhouse over HTTPS, which is not supported by the current version of r2d2-mysql.

After researching this topic, I found that ClickHouse has released its Rust library for use with ClickHouse.
https://github.com/ClickHouse/clickhouse-rs

If you don't have anything strong against it, I could start working on implementation of native support of their native library for clickhouse.

**Comment by hohmannr** [comment]
I would be also thrilled if we could use the native/http(s) protocol. I would be willing to take it as well. @Spar9a

**Comment by wangxiaoying** [comment]
> Hi @wangxiaoying, I think it could also be related to my problem.
> 
> My team and I using clickhouse over HTTPS, which is not supported by the current version of r2d2-mysql.
> 
> After researching this topic, I found that ClickHouse has released its Rust library for use with ClickHouse. https://github.com/ClickHouse/clickhouse-rs
> 
> If you don't have anything strong against it, I could start working on implementation of native support of their native library for clickhouse.

I think it looks good! Please feel free to open a PR on this and let me know if you need my help on anything.

**Comment by Spar9a** [comment]
@hohmannr Yeah, I'll take it and try to finish it in the short term because I see some good points in improving the existing services on my team.

@wangxiaoying Great, I am starting to work on it.  

I will be in touch regarding problems or questions!

**Comment by IlluvatarEru** [comment]
> @hohmannr Yeah, I'll take it and try to finish it in the short term because I see some good points in improving the existing services on my team.
> 
> @wangxiaoying Great, I am starting to work on it.
> 
> I will be in touch regarding problems or questions!

I think it would improve the lives of a lot of people, can't wait to see it merged
